### ISSUE TITLE
Infernal Invocation r21 DS4 and GLM source merge contract

### ISSUE THREAD

**Issue by voipmonitor** [voipmonitor/rtx6kpro#67]
# Infernal Invocation r21 Source Merge Contract

## Status

**Qualified for DeepSeek-V4-Flash-0731 TP2/DCP1 fixed probabilistic DSpark K5 serving:**

```text
image:          voipmonitor/vllm:infernal-invocation-vllmd6cf36a-b12xf6dc512-fi1ac6942-cu133-torch213-20260827-r21
registry digest: sha256:ed525dec1a4ac5cf7f19c7cf2fb29661389d71a29ff8de91aade8e6785e10291
image ID:       sha256:24f19364f0c6a991422bcb436a3e07ab52e66e0eb241aba0b9490e95476a8e3f
Docker source:  local-inference-lab/blackwell-llm-docker@cc2ac998e8f7b5f04d4271a79e6647b4debad3db
```

**Implemented but not branch-only:** the source locks reproduce the qualified trees from public base commits and named pull-request heads. They contain no local source patch or research overlay. Twenty-four vLLM pull requests and three B12X pull requests remain open. Seven additional vLLM heads in the integration lock are already merged into `dev/infernal-invocation`; the pinned base predates those merges.

**Qualified scope:** DeepSeek-V4-Flash-0731 completed model loading, FULL target/DSpark/DFlash graph capture, exact response sanity, and a 30-second C1 gate. GLM-5.2 source contracts are present in the integration tree, but r21 did not repeat a GLM model-load E2E gate.

The deployment specification is [DeepSeek-V4-Flash-0731 Infernal Invocation r21](https://github.com/local-inference-lab/rtx6kpro/blob/master/models/ds4dspark-infernal-invocation-r21.md).

## Source Identities

| Component | Canonical base | Qualified integration tree |
|---|---|---|
| vLLM | `dev/infernal-invocation@b5f995e73e6b7fe27c9927477e277a151ebcc9e9` | `d6cf36ae0dc30d48fd656a3c34a353ec62074922` |
| B12X | `master@a71c705f1c4710f59129562d26c73e70098e29de` | `f6dc512eb13ac2c09b2bf53656c704081af64361` |
| LMCache | `release/v0.5.2-glm52-dcp-base@a128b2e286ebb3556cb43124149e600ff99fe481` | `e045d729bc5c4c63a40e13d032f42923de97812f` |

The immutable locks are stored under `patches/releases/infernal-invocation-r21` in `local-inference-lab/blackwell-llm-docker@cc2ac99`. The image labels contain the complete pull-request list, patch digests, dependency revisions, and ABI versions.

## vLLM Merge Queue

Merge these open, non-draft pull requests into `dev/infernal-invocation` in the listed order:

1. [#285](https://github.com/local-inference-lab/vllm/pull/285) preserves resolved Hugging Face revisions across worker spawn.
2. [#286](https://github.com/local-inference-lab/vllm/pull/286) inherits target identity for in-checkpoint speculative drafts.
3. [#287](https://github.com/local-inference-lab/vllm/pull/287) defines DeepSeek V4 launch identity, graph capacity, and checkpoint contracts.
4. [#288](https://github.com/local-inference-lab/vllm/pull/288) enforces the B12X mHC input-shape contract for DeepSeek V4 MTP.
5. [#290](https://github.com/local-inference-lab/vllm/pull/290) profiles scheduler-reachable serving shapes before KV-cache sizing.
6. [#292](https://github.com/local-inference-lab/vllm/pull/292) enforces the canonical `b12x` package identifier.
7. [#295](https://github.com/local-inference-lab/vllm/pull/295) preserves XGrammar termination across token batches.
8. [#296](https://github.com/local-inference-lab/vllm/pull/296) absorbs duplicate DeepSeek V4 tool-call closers.
9. [#298](https://github.com/local-inference-lab/vllm/pull/298) requires request decode state before FULL CUDA-graph dispatch.
10. [#300](https://github.com/local-inference-lab/vllm/pull/300) loads projection-mixed MCG K3/K4/K5 routed experts and persists online K6 payloads.
11. [#301](https://github.com/local-inference-lab/vllm/pull/301) enforces GLM-5.2 B12X sparse-MLA runtime contracts.
12. [#302](https://github.com/local-inference-lab/vllm/pull/302) activates reasoning-aware structural tool grammars from token zero.
13. [#303](https://github.com/local-inference-lab/vllm/pull/303) preserves DeepSeek V4 FP8 quantization with an in-checkpoint DSpark draft.
14. [#304](https://github.com/local-inference-lab/vllm/pull/304) publishes native filesystem KV blocks without replacing visible destination inodes.
15. [#308](https://github.com/local-inference-lab/vllm/pull/308) zeroes recycled physical blocks for heterogeneous attention caches.
16. [#309](https://github.com/local-inference-lab/vllm/pull/309) defers MLA DCP workspace allocation until CUDA materialization.
17. [#320](https://github.com/local-inference-lab/vllm/pull/320) validates accepted speculative tokens against structured-output grammar before scheduler commit.
18. [#415](https://github.com/local-inference-lab/vllm/pull/415) preserves FULL CUDA-graph dispatch for DSpark capture-only contexts.
19. [#417](https://github.com/local-inference-lab/vllm/pull/417) parses the legacy direct DeepSeek V4 DSML tool-call form.
20. [#422](https://github.com/local-inference-lab/vllm/pull/422) restores failed external KV loads independently across heterogeneous cache groups.
21. [#476](https://github.com/local-inference-lab/vllm/pull/476) force-rejects grammar-invalid speculative suffixes and recovers tokens from the target distribution.
22. [#482](https://github.com/local-inference-lab/vllm/pull/482) repairs unaligned local prefix-cache restores for DeepSeek V4 compressed state.
23. [#483](https://github.com/local-inference-lab/vllm/pull/483) delegates TP2 CUDA-graph all-reduce routing to B12X.
24. [#485](https://github.com/local-inference-lab/vllm/pull/485) uses the public B12X DeepSeek attention namespaces.

The integration lock also contains these already merged heads because the pinned base commit predates their branch merges:

- [#423](https://github.com/local-inference-lab/vllm/pull/423) skips absent fused MTP indexer targets for split EXL3 projections.
- [#429](https://github.com/local-inference-lab/vllm/pull/429) keeps DeepSeek V4 indexer scoring inside breakable CUDA graphs.
- [#430](https://github.com/local-inference-lab/vllm/pull/430) sizes DeepSeek V4 C128A graph metadata from physical graph capacity.
- [#431](https://github.com/local-inference-lab/vllm/pull/431) enforces the active-row contract for DeepSeek V4 sparse top-k.
- [#432](https://github.com/local-inference-lab/vllm/pull/432) preserves GLM sparse-attention row semantics during prefill.
- [#433](https://github.com/local-inference-lab/vllm/pull/433) resets cached MRV2 logits-processing state for each request.
- [#434](https://github.com/local-inference-lab/vllm/pull/434) updates DeepSeek V4 sparse metadata without per-step host scalar extraction.

Pull request [#294](https://github.com/local-inference-lab/vllm/pull/294) is excluded because #476 implements the complete grammar-invalid speculative-token contract. Pull requests #289 and #293 are closed and superseded by #430 and #422 respectively.

## B12X Merge Queue

Merge these open, non-draft pull requests into `master` in order:

1. [#243](https://github.com/local-inference-lab/b12x/pull/243) executes FP16/BF16 K6/MCG decode with a native fused kernel.
2. [#246](https://github.com/local-inference-lab/b12x/pull/246) makes qualified TP2 CUDA-graph peer-push generation-safe and records reproducible transport evidence.
3. [#247](https://github.com/local-inference-lab/b12x/pull/247) bounds native W4A16 route execution by resident tensors.

All three heads are mergeable. The r21 image contains B12X #246 through commit `37dc7f8036f602adc1edac47464876d2a9c61d71` and passed its focused two-GPU CUDA-graph payload, storage-layout, and generation-wrap test in 9.61 seconds. The merge-ready #246 head `2a2baaf3979c1f47f8fd139aab5c6eeea35003e4` additionally rejects overlapping raw-pointer views while preserving transposed storage-dense views; its repository suite passed 230 tests with 2 skips, and the complete two-GPU CUDA torture suite passed 2 tests in 15.60 seconds.

## LMCache State

LMCache pull requests #7 through #17, #22, and #23 are present in `release/v0.5.2-glm52-dcp-base@a128b2e286ebb3556cb43124149e600ff99fe481`. The resulting tree is `e045d729bc5c4c63a40e13d032f42923de97812f`; the r21 LMCache integration patch is empty.

## Excluded Research

The source locks explicitly exclude:

- the external row-specific communication calibration program;
- the runtime FP8 dense-GEMM autotuner;
- deterministic split-K FP8 projection experiments;
- PDL launch experiments;
- Kimi-K3, FLA, and KVarN model-specific pull requests.

These changes are research-only because their measured variants did not establish a qualified end-to-end gain over the retained runtime or have a model-specific contract unrelated to DS4 and GLM.

## Qualification Evidence

The DS4 gate used TP2/DCP1 fixed probabilistic K5, `MAX_NUM_SEQS=8`, graph cap 48, `MAX_NUM_BATCHED_TOKENS=4096`, FP8 compressed MLA KV, and the `b12x-a8-dglin` backend on two 600 W RTX PRO 6000 Blackwell GPUs connected through one PCIe switch.

- The Docker source-lock, release, and Compose tests passed.
- The runtime contract passed with CUDA 13.3, PyTorch 2.13.0, NCCL 2.31.2, FlashInfer 0.6.18, CUTLASS DSL 4.6.2, XGrammar 0.2.5, and LMCache 0.5.2.
- InstantTensor `BUFFERED` model loading completed.
- Target, DSpark, and DFlash context-KV FULL CUDA graphs captured.
- Exact response sanity returned `r21 validation passed`.
- The 30-second C1 gate measured 177.66 aggregate tok/s, 68.45 target steps/s, 2.595 accepted tokens per verifier step, and 31.90% strict accepted/drafted.
- vLLM reported 1,249,424 aggregate compressed-MLA KV tokens under the qualified scheduler envelope.
- The r21 B12X integration tree passed exact TP2 BF16 payload, signed-zero, transposed storage-dense view, and generation-wrap graph replay coverage.

The r21 qualification repeated only the DS4 C1 E2E gate. TP2/TP4 concurrency, prefill, long-context, and persistent-cache evidence remains documented as historical evidence in the r19 serving study and is not relabeled as an r21 measurement.

## Completion Criteria

1. Merge every open pull request in the vLLM and B12X queues into its named canonical branch.
2. Recompose from the resulting branch heads with empty pull-request and `source_patches` lists.
3. Require the recomposed source trees to match the qualified integration trees, or document and qualify every intentional tree difference.
4. Repeat the Docker source-contract suite, DS4 FULL graph capture, C1 gate, strict-tool soak, persistent-cache restart tests, and GLM model-load gate.
5. Preserve the r21 registry digest as the rollback identity until the branch-only composition passes those gates.
